Added PRF docs (#307)
bitwarden/contributing-docs: docs/architecture/deep-dives/passkeys/implementations/relying-party/prf.md

# Passkeys for Decryption

When a user logs into their Bitwarden vault, two related but distinct operations take place -
authentication and decryption.

When using a master password, the password itself serves both of these purposes - authenticating the
user by verifying the password was entered correctly, and allowing decryption of your data through
the derivation of an encryption key from the password. For more detail on this, see the
[Bitwarden Security Whitepaper](https://bitwarden.com/help/bitwarden-security-white-paper/).

Passkeys inherently provide the mechanism for authentication. The `prf` extension introduced in the
WebAuthn [specification](https://w3c.github.io/webauthn/#prf-extension) provides the framework
necessary for passkeys to be used for decryption as well.

## What is PRF?

At the most abstract level, account encryption and decryption requires that the user's symmetric key

> be encrypted ("protected") with another key that is available only to the client. For master

Reviewer: question: do you think "client" might be confusing? I first thought of the WebAuthn Client, but in this context we are referring to the "client application"?

suggestions (non-blocking):
```suggestion
be encrypted ("protected") with another key that is available only to the client application. For master
```

```suggestion
be encrypted ("protected") with another key that is only available client-side. For master
```

```suggestion
be encrypted ("protected") with another key that is available only to the user. For master
```

```suggestion
be encrypted ("protected") with another key that is never shared with the server. For master
```

Author: I addressed this with [`4ab27ff` (#307)](https://github.com/bitwarden/contributing-docs/pull/307/commits/4ab27ff50f7a78620ea68d6f99691c481f24c263).

> password-based decryption, this is done with the master key, derived through a key derivation
> function from the master password.

Reviewer: suggestion: remove 1 instance of "master password" from the sentence

```suggestion
password-based decryption, the key is derived from the master password using a key derivation
function.
```

Author: I addressed this with [`4ab27ff` (#307)](https://github.com/bitwarden/contributing-docs/pull/307/commits/4ab27ff50f7a78620ea68d6f99691c481f24c263).

> In order to use a passkey for account decryption on login, we need such a key available on the
> passkey, with which we can encrypt and decrypt the user's symmetric key. Fortunately, the `prf`
> extension to the WebAuthn spec provides exactly this functionality.

Reviewer: suggestion: This paragraph seems to be tightly connected to the previous one, consider merging.

suggestion: Simplify. Remove "we" and use "client" which is used earlier in the paragraph
```suggestion
In order to use a passkey for account decryption, the client needs to able to derive such a key from the passkey. Fortunately, the `prf` extension provides exactly this functionality.
```

Author: I addressed this with [4ab27ff (#307)](https://github.com/bitwarden/contributing-docs/pull/307/commits/4ab27ff50f7a78620ea68d6f99691c481f24c263).

> The `prf` extension, which is built upon the `hmac-secret` CTAP2 extension, provides the

Reviewer: issue: while it is true that the current implementations of `prf` are built upon `hmac-secret`, this might not always be true. `prf` is a client extension and `hmac-secret` is a CTAP2 authenticator extension. Passkey providers (like e.g. Apple and ourselves) that don't implement CTAP2 might implement `prf` differently.

Author: Thank you!  This was very interesting to dig into, and I've expanded the section and included a callout to distinguish the two extensions in [4ab27ff (#307)](https://github.com/bitwarden/contributing-docs/pull/307/commits/4ab27ff50f7a78620ea68d6f99691c481f24c263).

specification by which credential assertion can return the output from a pseudo-random function

> (hence "PRF") that is unique to each credential on the authenticator. The Bitwarden relying party

Reviewer: > a pseudo-random function that is unique to each credential on the authenticator

nit: it's not the function that is unique, but the secret it hashes together with the salt provided by the RP

Author: Just to make 100% sure - the secret itself is also unique to each credential, correct? Even without the salt?

Reviewer: Yes, the secret is unique! We're using the same salt across all credentials so the only thing that differs is the secret

When a passkey is registered for use with Bitwarden, we interrogate the attestation response to see
if PRF is supported. If so, we allow the user to choose to enable encryption.

![Registering a passkey with PRF](image.png)

If the user chooses to enable encryption, we request an assertion from the authenticator, which will
provide the key. (This is why you may be prompted multiple times to provide user verification when
registering a PRF-capable passkey.)

The 32-byte key returned from the authenticator is then stretched to a 64-byte two-part key using
HKDF. The first 32 bytes will serve as the AES-256 encryption key and the second 32 bytes will serve
as a message authentication code (MAC).

Once we have this encryption key, we can encrypt the user's symmetric key as follows:

    1. Create an RSA public/private key pair, specific to this credential.
    2. Encrypt the RSA private key with the encryption key.
    3. Encrypt the user's symmetric key with the RSA public key.

The PRF-encrypted private key, the public key-encrypted encryption key, and the RSA public key are
persisted to the server and associated with the registered passkey. Note that the PRF key does

> **not** leave the client at any point, to ensure zero-knowledge encryption.

Author: I addressed this with [4ab27ff (#307)](https://github.com/bitwarden/contributing-docs/pull/307/commits/4ab27ff50f7a78620ea68d6f99691c481f24c263).

:::tip Why an RSA key pair?

Put simply: key rotation. In order to support
[encryption key rotation](https://bitwarden.com/help/account-encryption-key/#rotate-your-encryption-key),
a user must be able to re-encrypt the symmetric key for all registered passkeys without gaining

> access to the PRF keys for each registered credential. The use of a public/private keypair allows a

Reviewer: nit: key pair is written in two words earlier in the document

```suggestion
access to the PRF keys for each registered credential. The use of a public/private key pair allows a
```

Author: I addressed this with [4ab27ff (#307)](https://github.com/bitwarden/contributing-docs/pull/307/commits/4ab27ff50f7a78620ea68d6f99691c481f24c263).

> access to the PRF keys for each registered credential. The use of a public/private keypair allows a

Reviewer: > a user **must** ... re-encrypt the symmetric key ... without gaining access to ... each registered credential.

nit: "must" is a strong word. There's nothing that inherently dictates that. This is more about a choice we've made to simplify the user experience.

suggestion (with room for improvement):
In order to support [encryption key rotation](https://bitwarden.com/help/account-encryption-key/#rotate-your-encryption-key), a user must be able to re-encrypt the symmetric key for all registered passkeys. If the PRF key was used to directly encrypt the user's symmetric key then this operation would require gaining access to the PRF keys for each registered credential. The use of a public/private keypair allows the user to re-encrypt the new symmetric key with the public key, with the assurance that only the holder of the private key can decrypt it, and without having to provide access to the PRF key for each registered credential.

Author: I think your suggestion sounds great!  I've updated the copy in [4ab27ff (#307)](https://github.com/bitwarden/contributing-docs/pull/307/commits/4ab27ff50f7a78620ea68d6f99691c481f24c263).

user to re-encrypt the new symmetric key with the public key, with the assurance that only the
holder of the private key can decrypt it.

:::

### During passkey login

> When a user chooses to authenticate with a passkey, we request an assertion from the authenticator.

Reviewer: nit: don't use "we"
```suggestion
When a user chooses to authenticate with a passkey, an assertion is requested from the authenticator.
```

Author: I addressed this with [4ab27ff (#307)](https://github.com/bitwarden/contributing-docs/pull/307/commits/4ab27ff50f7a78620ea68d6f99691c481f24c263).

If the passkey is registered for encryption, and the
[platform supports it](#what-if-i-am-prompted-for-my-master-password-when-using-a-prf-enabled-passkey),
the Bitwarden client will receive the PRF key in the assertion response.

The client will then submit the assertion (without the PRF key) to the server to perform
authentication.

If the server is able to validate the assertion, the user is authenticated, and as a part of the

> authentication response the PRF-encrypted key data is returned to the client:

Reviewer: ```suggestion
authentication response the following PRF-encrypted key data is returned to the client:
```

Author: I addressed this with [4ab27ff (#307)](https://github.com/bitwarden/contributing-docs/pull/307/commits/4ab27ff50f7a78620ea68d6f99691c481f24c263).

    - The PRF-encrypted RSA private key
    - The RSA public key-encrypted user symmetric key

In order to decrypt the vault data:

    1. The PRF key from the credential is stretched with HKDF.
    2. The resultant 32-bit encryption key is used to decrypt the RSA private key.
    3. The RSA private key is used to decrypt the user's symmetric key.
    4. The symmetric key is used for vault decryption.

## Tips for using your passkey for decryption

### What if I am prompted for my master password when using a PRF-enabled passkey?

Support for PRF varies greatly across authenticators and platforms. Only when all members of the

> WebAuthn attestation ceremony support PRF can Bitwarden obtain the key necessary for decryption. If

Author: I addressed this with [4ab27ff (#307)](https://github.com/bitwarden/contributing-docs/pull/307/commits/4ab27ff50f7a78620ea68d6f99691c481f24c263).
